### ISSUE TITLE
[RFC]: rename (or redefine?) math.base.assert.isInteger

### ISSUE THREAD

**Issue by lightmare** [stdlib-js/stdlib#469]
### Description

I got confused when reading some code using `isInteger`, because there are two different functions of that name:
- [assert.isInteger](https://stdlib.io/docs/api/latest/@stdlib/assert/is-integer), which returns `false` for infinities (like built-in `Number.isInteger`)
- [math.base.assert.isInteger](https://stdlib.io/docs/api/latest/@stdlib/math/base/assert/is-integer), which returns `true` for infinities

The difference is clearly explained in the documentation, which is great, but I still think the naming is unfortunate.
I suggest to change `math.base.assert.isInteger`:
- either rename to `isIntegerOrInfinity`
- or make it return `false` for infinities

---

While on this topic, wouldn't it make more sense for `assert.isInteger` — instead of wrapping "the other isInteger" — to use `Number.isInteger` if available, and reimplement that if not? Like:
```js
var isInteger = Number.isInteger || function isInteger( value ) {
  return x === x - x % 1
}
```

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

**Comment by kgryte** [comment]
Hey @lightmare! Thanks for opening this issue.

`@stdlib/math/base/assert/is-integer` is a "base" implementation, and what is meant by that is that it eschews certain checks and assumes particular aspects of input values without validating that those assumptions are valid.

We use "base" implementations throughout the project in those contexts where we have knowledge about value characteristics (e.g., we might know that by the time `isInteger` is called with `v` that `v` cannot be infinity) and we'd like to avoid the overhead of the additional checks and constraints imposed by non-"base" implementations, such as `@stdlib/assert/is-integer`.

Re: `Number.isInteger`. We often avoid relying on built-ins, especially those which may not have consistent support across all the environments we support (e.g., Node v0.10, IE, etc). Often, as in this case, we're also interested in the implementation; here, we're wanting to avoid modulo arithmetic and/or division, preferring instead to `floor`, which can be implemented without division (e.g., see [musl](https://git.musl-libc.org/cgit/musl/tree/src/math/floor.c)). According to the [spec](https://www.ecma-international.org/wp-content/uploads/ECMA-262_12th_edition_june_2021.pdf) (see 7.2.6), native implementations must also compute the absolute value (not clear why), so we avoid unnecessary computation by rolling our own.

In short, while I can see where multiple `isInteger` can be confusing, I don't think we're likely to rename or change the implementation. The reason for the former is that we commonly shadow package names in "base" sub-namespaces, often because we want more performant variants of related non-"base" packages for use internally. The reason for the latter is that we heavily use `@stdlib/math/base/is-integer` throughout the project internally and rely on it assuming finite input in order to squeeze out a bit more performance.

Hopefully that answers your questions/concerns. 😅

**Comment by lightmare** [comment]
Wow, that explains a lot. Though putting it all together leads me back to `Number.isInteger`.

> According to the [spec](https://www.ecma-international.org/wp-content/uploads/ECMA-262_12th_edition_june_2021.pdf) (see 7.2.6), native implementations must also compute the absolute value (not clear why), so we avoid unnecessary computation by rolling our own.

The [operation in question](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-isintegralnumber)

> — 3. If floor(abs(ℝ(argument))) ≠ abs(ℝ(argument)), return false.

is defined in terms of [mathematical operations](https://tc39.es/ecma262/multipage/notational-conventions.html#sec-mathematical-operations) on real numbers. I guess the `abs` in there is only a hint that when implementing this in binary floating point, you can ignore the sign bit. Skipping it won't change the result.

And the wording used for real [floor](https://tc39.es/ecma262/multipage/notational-conventions.html#eqn-floor):

> — The mathematical function floor(x) produces the largest integer (closest to +∞) that is not larger than x.

... is nearly identical to the wording used for [Math.floor](https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-math.floor):

> — 5. Return the greatest (closest to +∞) integral Number value that is not greater than n.

---

Of course one cannot extrapolate implementation performance from reading the spec. It just seems counter-intuitive to me to avoid using `Number.isInteger` when it's available. I don't see how `Math.floor` + compare could ever be faster.

**Comment by kgryte** [comment]
Luckily, we have some benchmarks for assessing the difference, with the caveat that perf will be dependent on the host machine, Node.js version (and thus V8 version), etc. Here are my results:

```text
$ make benchmark BENCHMARKS_FILTER=.*/assert/is-integer/.*

Running benchmark: /@stdlib/assert/is-integer/benchmark/benchmark.js
TAP version 13
# @stdlib/assert/is-integer::primitives,true
  ---
  iterations: 100000000
  elapsed: 0.459187457
  rate: 217775983.37142733
  ...
ok 1 benchmark finished
# @stdlib/assert/is-integer::primitives,true
  ---
  iterations: 100000000
  elapsed: 0.481124233
  rate: 207846525.16141292
  ...
ok 2 benchmark finished
# @stdlib/assert/is-integer::primitives,true
  ---
  iterations: 100000000
  elapsed: 0.479482899
  rate: 208558011.5757163
  ...
ok 3 benchmark finished
# @stdlib/assert/is-integer::primitives,false
  ---
  iterations: 100000
  elapsed: 0.289117907
  rate: 345879.64833323174
  ...
ok 4 benchmark finished
# @stdlib/assert/is-integer::primitives,false
  ---
  iterations: 100000
  elapsed: 0.254431696
  rate: 393032.79258099984
  ...
ok 5 benchmark finished
# @stdlib/assert/is-integer::primitives,false
  ---
  iterations: 100000
  elapsed: 0.325634142
  rate: 307093.10573459463
  ...
ok 6 benchmark finished
# @stdlib/assert/is-integer::objects,true
  ---
  iterations: 10000000
  elapsed: 0.198526332
  rate: 50371151.77245102
  ...
ok 7 benchmark finished
# @stdlib/assert/is-integer::objects,true
  ---
  iterations: 10000000
  elapsed: 0.252158161
  rate: 39657649.62887717
  ...
ok 8 benchmark finished
# @stdlib/assert/is-integer::objects,true
  ---
  iterations: 10000000
  elapsed: 0.195392584
  rate: 51179015.06435884
  ...
ok 9 benchmark finished
# @stdlib/assert/is-integer::objects,false
  ---
  iterations: 100000
  elapsed: 1.030203845
  rate: 97068.16809638291
  ...
ok 10 benchmark finished
# @stdlib/assert/is-integer::objects,false
  ---
  iterations: 100000
  elapsed: 1.090669978
  rate: 91686.76319795061
  ...
ok 11 benchmark finished
# @stdlib/assert/is-integer::objects,false
  ---
  iterations: 100000
  elapsed: 1.034096595
  rate: 96702.76498686275
  ...
ok 12 benchmark finished
# @stdlib/assert/is-integer::primitives,true:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.484704454
  rate: 206311287.57896662
  ...
ok 13 benchmark finished
# @stdlib/assert/is-integer::primitives,true:isPrimitive
  ---
  iterations: 100000000
  elapsed: 1.378592569
  rate: 72537747.73538621
  ...
ok 14 benchmark finished
# @stdlib/assert/is-integer::primitives,true:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.488753104
  rate: 204602281.1550267
  ...
ok 15 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.705010802
  rate: 141841798.33318356
  ...
ok 16 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 1.693874602
  rate: 59036247.35970863
  ...
ok 17 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.701501167
  rate: 142551437.8367385
  ...
ok 18 benchmark finished
# @stdlib/assert/is-integer::objects,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.47419496
  rate: 210883725.96790147
  ...
ok 19 benchmark finished
# @stdlib/assert/is-integer::objects,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 1.512978385
  rate: 66094797.51424208
  ...
ok 20 benchmark finished
# @stdlib/assert/is-integer::objects,false:isPrimitive
  ---
  iterations: 100000000
  elapsed: 0.502634841
  rate: 198951588.39575946
  ...
ok 21 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isObject
  ---
  iterations: 100000
  elapsed: 0.197234001
  rate: 507011.9730522528
  ...
ok 22 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isObject
  ---
  iterations: 100000
  elapsed: 0.177974986
  rate: 561876.7122701164
  ...
ok 23 benchmark finished
# @stdlib/assert/is-integer::primitives,false:isObject
  ---
  iterations: 100000
  elapsed: 0.18170964
  rate: 550328.5351289012
  ...
ok 24 benchmark finished
# @stdlib/assert/is-integer::objects,true:isObject
  ---
  iterations: 10000000
  elapsed: 0.182287115
  rate: 54858512.627181575
  ...
ok 25 benchmark finished
# @stdlib/assert/is-integer::objects,true:isObject
  ---
  iterations: 10000000
  elapsed: 0.18181103
  rate: 55002163.51010167
  ...
ok 26 benchmark finished
# @stdlib/assert/is-integer::objects,true:isObject
  ---
  iterations: 10000000
  elapsed: 0.187180854
  rate: 53424267.42000012
  ...
ok 27 benchmark finished
# @stdlib/assert/is-integer::objects,false:isObject
  ---
  iterations: 100000
  elapsed: 1.030341546
  rate: 97055.19532646313
  ...
ok 28 benchmark finished
# @stdlib/assert/is-integer::objects,false:isObject
  ---
  iterations: 100000
  elapsed: 0.970946112
  rate: 102992.32754948195
  ...
ok 29 benchmark finished
# @stdlib/assert/is-integer::objects,false:isObject
  ---
  iterations: 100000
  elapsed: 0.962919034
  rate: 103850.89137203616
  ...
ok 30 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,true
  ---
  iterations: 100000000
  elapsed: 0.42710983
  rate: 234131815.6971475
  ...
ok 31 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,true
  ---
  iterations: 100000000
  elapsed: 1.2610018250000001
  rate: 79302026.3868373
  ...
ok 32 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,true
  ---
  iterations: 100000000
  elapsed: 0.413350882
  rate: 241925212.58488572
  ...
ok 33 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,false
  ---
  iterations: 100000000
  elapsed: 0.664010764
  rate: 150599968.28605628
  ...
ok 34 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,false
  ---
  iterations: 100000000
  elapsed: 1.566836455
  rate: 63822870.396515
  ...
ok 35 benchmark finished
# @stdlib/assert/is-integer::builtin,primitives,false
  ---
  iterations: 100000000
  elapsed: 0.873182742
  rate: 114523564.41556881
  ...
ok 36 benchmark finished
#
1..36
# total 36
# pass  36
#
# ok

Running benchmark: /@stdlib/math/base/assert/is-integer/benchmark/benchmark.js
TAP version 13
# @stdlib/math/base/assert/is-integer::true
  ---
  iterations: 100000000
  elapsed: 0.375061884
  rate: 266622667.5275806
  ...
ok 1 benchmark finished
# @stdlib/math/base/assert/is-integer::true
  ---
  iterations: 100000000
  elapsed: 0.386656315
  rate: 258627613.51770502
  ...
ok 2 benchmark finished
# @stdlib/math/base/assert/is-integer::true
  ---
  iterations: 100000000
  elapsed: 0.392391342
  rate: 254847620.97528645
  ...
ok 3 benchmark finished
# @stdlib/math/base/assert/is-integer::false
  ---
  iterations: 100000000
  elapsed: 0.442580432
  rate: 225947630.68964604
  ...
ok 4 benchmark finished
# @stdlib/math/base/assert/is-integer::false
  ---
  iterations: 100000000
  elapsed: 1.283130782
  rate: 77934378.47709587
  ...
ok 5 benchmark finished
# @stdlib/math/base/assert/is-integer::false
  ---
  iterations: 100000000
  elapsed: 0.415609983
  rate: 240610197.27719098
  ...
ok 6 benchmark finished
#
1..6
# total 6
# pass  6
#
# ok
```

How to read the above:

-   `# @stdlib/assert/is-integer::primitives,true`: this is the benchmark name, which displays the package name along with facets following `::`, and, if present, `:` giving the method name, along with any explicit parameter/option values. Here, the benchmark is for the package `@stdlib/assert/is-integer` where the main export is provided primitive values which should always evaluate to `true` following the `isInteger` operation.
-    The YAML following the benchmark name contains the benchmark results for a single run. Each benchmark is run in triplicate. Iterations gives the number of benchmark iterations (i.e., each time an operation is performed). Elapsed gives the total time (in seconds) required to perform all iterations. Rate gives the number of iterations per second, where higher is better.
-   In general, only the highest rate matters.

Takeaways:

-   mainly interested in `:isPrimitive` benchmarks, as these measure what is ostensibly the equivalent of `Number.isInteger`. In general, the `stdlib` implementation is effectively equivalent to the built-in.
-   the above also includes `@stdlib/math/base/assert/is-integer` and what is important to note here is that is more consistent across the board for operations resulting in either `true` or `false`. Why is this noteworthy? Because that is precisely what we want when used internally (e.g., in transcendental functions) when we know something about the nature of the values to test.

One other addendum is that we do lose out on possible peak performance when a value is considered by V8 to be a small integer (a.k.a., an smi). When iterating over an array of only small integers, `Number.isInteger` is demonstrably faster, which may be expected when V8 knows that it can avoid calling `Number.isInteger` altogether. In this case, V8 is privy to knowledge we don't have when implementing this in userland--namely, how a value is stored in memory.

So in short, certainly `@stdlib/math/base/assert/is-integer` is not likely to change, as it exhibits perf behavior we find desirable--namely, running equivalently fast for both ints and non-ints.

For `@stdlib/assert/is-integer` (and, in particular, `isPrimitive`), I'd consider it a toss-up in the general sense, with more potential upside for using the built-in for certain special cases (smis). However, this upside comes with an asterisk, in that the perf profile may be specific to V8 and vary across runtimes. In which case, I still lean against changing our current implementation as the perf profile is, I'd argue, relatively predictable and likely to be consistent across engines (e.g., guards against a runtime using `%` to check for an integer, etc).

All this stated, we have discussed figuring out a way for users to bundle the project and swap out functionality which has a built-in behavioral equivalent. This is mainly for the browser use case, where bundle size concerns find greater weight. However, we're only at the ideation stage on this front and have yet to write any code to make this a reality.

**Comment by kgryte** [comment]
As this RFC is unlikely to be accepted, will close out...
